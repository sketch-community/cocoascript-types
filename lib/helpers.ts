import { NAMESPACE } from './constant';
import { shim } from './shim';
import { extend } from './extend';

export function capitalize(str: string) {
  return str[0].toUpperCase() + str.slice(1, str.length);
}

export function normalizeType(
  type: string,
  options: {
    extend?: boolean;
    withNamespace?: boolean;
  } = {
    extend: true,
    withNamespace: false,
  }
) {
  let finalType = type;
  if (options.extend && extend[finalType]) {
    return extend[finalType];
  }
  if (shim[finalType]) {
    return shim[finalType];
  }
  if (options.withNamespace) {
    return `${NAMESPACE}.${finalType}`;
  }
  return finalType;
}

export function getTokens(doc: any) {
  return doc.primaryContentSections?.[0]?.declarations?.[0]?.tokens || [];
}
