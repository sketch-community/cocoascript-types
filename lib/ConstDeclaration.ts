import { CodeGenerator } from './CodeGenerator';
import { normalizeType } from './helpers';
import { ParseError } from './ParseError';
import { Token } from './types';

export class ConstDeclaration {
  constructor(public id: string, public name: string, public type: string) {}

  static initFromTokens(id: string, tokens: Token[]) {
    let index = 0;
    let name;
    let type;
    while (index < tokens.length) {
      if (tokens[index].kind === 'typeIdentifier' || (tokens[index].kind === 'keyword' && tokens[index].text !== 'const')) {
        type = tokens[index].text;
      }
      if (tokens[index].kind === 'identifier') {
        name = tokens[index].text;
      }
      index += 1;
    }
    if (!name || !type) {
      throw new ParseError(id);
    }
    return new this(id, name, type);
  }

  generate() {
    const code = new CodeGenerator();
    code.appendLine(`// ${this.id}`);
    code.appendLine(`declare const ${this.name}: ${normalizeType(this.type)};`)
    code.appendLine();
    return code.toString();
  }
}
