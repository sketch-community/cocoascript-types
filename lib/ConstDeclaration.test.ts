import { ConstDeclaration } from './ConstDeclaration';
import { getTokens } from './helpers';

describe('ConstDeclaration', () => {
  it('simple const', () => {
    const doc = require('../documentation/appkit/nspasteboardtypestring.json')
    const decl = ConstDeclaration.initFromTokens('fakeid', getTokens(doc));
    expect(decl.generate()).toMatchSnapshot();
  });

  it('type before const', () => {
    const doc = require('../documentation/appkit/nsworkspacevolumeurlkey.json')
    const decl = ConstDeclaration.initFromTokens('fakeid', getTokens(doc));
    expect(decl.generate()).toMatchSnapshot();
  });
});
