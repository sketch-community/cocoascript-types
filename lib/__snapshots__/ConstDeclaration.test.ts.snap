// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConstDeclaration simple const 1`] = `
"declare const NSPasteboardTypeString: cocoascript.NSPasteboardType;
"
`;
